### PR TITLE
CJ4: 0.4.1 prerelease

### LINKED ISSUE #163: CJ4 Oscillation in flight
CJ4 mod 0.3.1 version oscillation in all phases of flight. May be at lighter loads did only 2 quick flights. with no pax, 1 pilot, and a large amount of fuel. Plane oscillates on autopilot in climb, cruise and descent, up and down pitching like a roller coaster. Wind at higher levels around 60kts tailwind.

### LINKED ISSUE #183: CJ4 v0.3.1: Use Oxygen mask at above FL440
Use Oxygen mask cation at above FL440
All button except the flight control related are untouched (cabin pressure dump)
Oxygen mask

### LINKED ISSUE #219: CJ4: MFD not centered on aircraft
The MFD does not center on the aircraft if going from upper menu OFF to FMS TEXT, CHECKLIST or PASSENGER BRIEF. Going to the SYSTEM page does correctly center the MFD and subsequently going from the SYSTEM page to one of the other three keeps the MFD centered correctly. See the video below:

https://youtu.be/6NwEZrIMFj8

On an unrelated note: any idea what happened to the engines near the end of my video? They seemed to overspeed twice while no input was given?

### LINKED ISSUE #220: CJ4: Legs and Progress page random (dis)appearance of WPTs
This one may be a bit difficult to reproduce. I will add more detail to this issue if I come across this more often. On final approach @EGNM in the below video, not all my final approach waypoints show up in the LEGS page. Going to the PROGRESS page does show them correctly. When I go back to the LEGS page, a past waypoint (DME09) suddenly appends to the already displayed legs. Back in the PROGRESS page now everything is gone. Hard to follow I suppose, see what I mean in the video below:

Version used is 0.4.0 release

https://youtu.be/fwA_erS_17M

### LINKED ISSUE #221: CJ4: Takeoff V-speeds do not display on PFD after first flight
After completing a flight, having successfully set both takeoff and approach V-speeds through the FMS > Send button, takeoff V-speeds display as -1. It is impossible to set them using the same procedure for a new flight plan. Calculating the V-speeds in the takeoff page for your new departure runway and pressing 'Send' does not do anything. Setting the speeds manually also does not make the V-speed bugs appear. Approach V-speeds are totally fine.

To reproduce:
- Complete a flight and set-up your next flight in the FMS. 

Version used is 0.4.0 release

Screenshot (307)

### LINKED ISSUE #222: CJ4: Aircraft Moving off When Parke Brake is released (without the engines running)
Hi

Thank you for this MOD its great, the work you guy’s have done is truly fantastic.

I have an issue with the latest version, where if I release the parking brake without having the engines running the aircraft takes off across the ground gathering speed as it goes (I crashed into a terminal 😀 my first crash, so crash damage works) . I then tried the same thing with the engines running and it did not do it this time. Not a major problem, as I can obviously just get the engines running before releasing park brake and getting a pushback.

I did remove all MODS from the community folder and test it again and the problem was still there.

Just thought I would raise it, in case anybody else notices the same.

### LINKED ISSUE #226: CJ4: FPLN does not add the complete airway
Starting from Fix NIK in Belgium: added Fix LAM in Great Britain via airway L610
Did not get the whole route but just a few fixes up until BULAM, which is on the airway L610 to LAM

Maybe because LAM is a sub string of BULAM?

### LINKED ISSUE #232: CJ4: Progress page showing wrong distance to destination
This issue is closely related to #220 and the discussion we had there. 

The indicated distance to the final waypoint on the progress page is wrong. The flight is as follows:

EKYT/26R VEPIP ADAXA LASMA EVLA/24 (ILS24 trans LAROKSO app)

Before reaching the first enroute waypoint (FMS TEXT is correct here)
Screenshot (313)

When I got close to VEPIP, the indication was almost correct
Screenshot (314)

However just after passing VEPIP, the PROGRESS page displayed a completely wrong distance once again:
Screenshot (315)

### LINKED ISSUE #241: CJ4 Weight seem to be incorrectly calculated
Hi, 

It seems like weight is incorrectly calculated in the CJ4 FMS.. 
I compare it to the calculation inside FS2020 and also in the flight planner I use (ForeFlight)..
It should be as it states in the payload manager which you can see in the attached picture but for some reason it gets waay to high everytime even though I add the correct cargo lbs etc in the FMS.

weight

### LINKED ISSUE #242: CJ4: Constant warning with PITOT / STATIC Heat on
Hi,

This issue I didn't have before the 0.4 update but everytime I turn on the PITOT / STATIC Heat I get the yellow PT Heat warning L/R..

Best Regards
Cisco

### PR DESCRIPTION
**Pull request by dga711**
Fix Bugs!

TODO
Things we don't fix here we should create issues for after release

- [x] **Changelog**
- [x] Fix direct to IAF from LEGS / closes #220 
- [x] Fix approach waypoint display in FMS Text / closes #220 
- [x] Fix distances in PROG/MFD / closes #232
- [x] Actually show errors when user inputs invalid waypoints 
- [x] Smoother AP / closes #163 
- [x] Add Magenta to TO fix on PROG
- [x] Reverse CROSSWIND L/R for tailwind situations on PROG2
- [x] Flight Dynamics set elevator_trim_effectiveness back to 1.5
- [x] Fix engines off thrust / closes #222 
- [x] Fix MFD plane on map not centered / closes #219 
- [x] Fix VSpeed -1 / closes #221
- [x] Fix approach waypoint display in LEGS / closes #220
- [x] Fix indexof when adding airways / closes #226 / credit @tscharlii for it
- [x] Added input filters for QNH on takeoff and landing perf pages
- [x] Fixed PITOT, OXY MASK and PARKING BRAKE CAS Messages / closes #242 / closes #183  
- [x] Added protection to Takeoff Ref and Approach Ref pages for missing departure or arrival runway
- [x] Fix Perf Init weight issues / closes #241 

**Audio** (still on another branch)
- [x] Need Chime when playing seatbelt announcement
- [x] Chimes are played when setting battery master to on
- [x] Need to set States correctly for the groups so sounds dont stop when going external view and back in. See https://discordapp.com/channels/750764704175226992/764122525323362325/764904295618969600
- [x] Add passenger briefing announcements
- [x] Restrict only one announcement to play at once
- [x] Improvements to climate control knob sound
- [x] Fix passenger sounds stopping when switching to outside view
- [x] Fix chime playing when battery is turned on

### CODE REVIEW

**Review by tscharlii — CHANGES_REQUESTED** [review]
Regarding 5579a87f597c6e7b30fb7c4f7b24f7a61c1fa09c

A simple equality check does not work. Note the fixed length format of airway.icaos. It's not just a list of identifiers, but contains more information.
Index 0: Type (A=Airport, V=VOR, W or I = Waypoint or Intersection, etc.)
Index 1 and following: I assume that's the region
Index 7, 8, 9, 10, 11: Identifier, padded with spaces for shorter than 5 character waypoints
```
airway.icaos = ["VEB    NIK  ","WEB    DENUT","WEB    LUMEN","WEB    BULAM","WEB    DIBLI","WEG    RAPIX","WEG    SUNUP","WEG    TEBRA","WEG    KOPUL","WEG    UNSAD","WEG    NILON","WEG    GILDA","VEG    LAM  ","WEG    HEMEL","WLT    EKSEN","WLT    EKMIM","WLT    UPKUL","WLT    EMKUD","WLT    OSGED","WLT    PESOX","WLT    SIVKU","WLT    APSER","WLT    ETAMP","WLT    EMGIM","VLT    YAA  ","VLT    IST  ","WLT    FENER","WLB    VADEN","WLY    UTEKA","WLY    DONIV","WLY    PELOV","WLY    TISAK","WLY    TONDO","WED    RENKA","WED    GONBA","WED    STAUB","WED    MAMOR","WED    UNKUL","WED    UPALA","WED    INBED","WED    AMOSA","WED    IBAGA","WED    BARSU","WED    KOMIB","WED    TUSOS","WED    BIBEG"]
```
With the following code in line 448, I was able to add the NIK -> L610 -> LAM airway
```
let lastWaypointIcao = airway.icaos.find(icao => icao.substring(7, 12) === lastWaypointIdent.padEnd(5, " "));
```

**Review by tscharlii — CHANGES_REQUESTED** [review]
Duplicate comment. Removed.

**Comment by dga711** [comment]
> Regarding [5579a87]
@tscharlii i see. i switched up icao and ident and didnt properly look into it because it was late ^^
i will check again and lookinto the vars while debugging. thanks for the heads up.

**Review by cwburnett — APPROVED** [review]
